### ISSUE TITLE
[Request] Add a method to use a function as an on use macro

### ISSUE THREAD

**Issue by tposney** [tposney/midi-qol#1203]
In GitLab by @chrisk123999 on Mar 19, 2023, 11:01

This would allow me to skip the need to include an item macro to pass along the midi-qol args.
Ideally being able to do this for `flags.midi-qol.onUseMacroName` too would be nice.  For context this would help me simplify the process of calling macros included in my module "Chris's Premades."

**Comment by tposney** [comment]
You can reference a macro compendium or world macro via a Compendium UUID or just a world macro name. Is that enough?

so flags.midi-qol.onUseMacroName CUSTOM echo, preItemRoll will call the world macro "echo" when an item is rolled.

**Comment by tposney** [comment]
In GitLab by @chrisk123999 on Mar 19, 2023, 22:00

Not really, I would just be doing the same thing.  Having to make a world or compendium macro pass along the args.

I'm trying to avoid having to do this as an item macro:
![Capture](https://my-gitlab-bucket.s3.amazonaws.com/640711674/21cf6e6299b309b4f51d3fa1464338189d3fdb2aa2ee7865102c077d73f70476/Capture.PNG)

The request is to be able to just directly call my exported function directly in this field:
![Capture](https://my-gitlab-bucket.s3.amazonaws.com/640711674/edd1199d1c441258c6c9af053e4e94ef23c3ac5f2a29ae3f3ec5a5445ff89444/Capture.PNG)

**Comment by tposney** [comment]
So what arguments would you want passed? I'm assuming that I'd call 
```js
chrisPremades.macro.callLightning.bind(this)()
```
the alternative (which I probably prefer) is 
```js
chrisPremades.macro.callLightning.bind(this)({ speaker, actor, token, character, item, args })
```

"this" would be the current workflow (if any).

**Comment by tposney** [comment]
I've implemented function. which can be used wherever a macro can be specified in midi-qol (but not DAE so macro.execute/macro.itemMacro don't support it - yet).

example
flags.midi-qol.onUseMacroName CUSTOM function.MidiQOL.log, preItemRoll displays 
```js
actor: Actor5e {name: 'Cassie', type: 'character', img: 'uploads//Cassie.Avatar.YrcI32qT9wVYREGv.png?t=1572480890785', system: CharacterData, #valid: true, …}
args: [{…}]
character: null
item: Item5e {name: 'Dagger', type: 'weapon', img: 'icons/weapons/daggers/dagger-jeweled-purple.webp', system: WeaponData, #valid: true, …}
speaker: {scene: 'QPG20BLsHJ7eug0j', token: 'sN1fjmodfvY9PeUG', actor: '8g5dIYAFv2iejVSV', alias: 'Cassie'}
token: Token5e {#controlled: false, #hover: false, #initialize: ƒ, #animateFrame: ƒ, #getMovementSource: ƒ, …}
```
and inside the function this refers to the workflow.

The function has to be accessible in the global scope for this to work.
Will go out in 10.0.35. I'm going to be away for 2 weeks so 10.0.35 won't go out until I'm back.

**Comment by tposney** [comment]
if you want to test it change callMacro in workflow.js to look like this
```js
  async callMacro(item, macroName: string, macroData: any): Promise<damageBonusMacroResult | any> {
    const name = macroName?.trim();
    // var item;
    if (!name) return undefined;
    let macroCommand;
    try {
      if (name.startsWith("function.")) {
        macroCommand = `${name.replace("function.", "").trim()}.bind(this)({ speaker, actor, token, character, item, args })`;
      } else if (name.startsWith(MQItemMacroLabel)) { // special short circuit eval for itemMacro since it can be execute as GM
```

**Comment by tposney** [comment]
Turns out I had another priority issue to fix, so this snuck in

**Comment by tposney** [comment]
In GitLab by @chrisk123999 on Mar 30, 2023, 06:08

Just a heads up these aren't being called async.  My current fix is to use "function.await myFunction" but it should likely be fixed on this end.

**Comment by tposney** [comment]
fixed in 10.0.36
